system/mp/test/mptest.cpp

```diff
@@ -521,127 +523,214 @@ void MultiTest(ICommunicator *_comm)
+            n1++;
+            if (n1 == numranks)
+                n1 = 0;
+            if (n1 == i)
+                n1++;
+            if (n1 == numranks)
+                n1 = 0;
+            na[i] = n1;
+            pa[na[i]] = i;
         }
+
+        next = na[myrank];
+        prev = pa[myrank];
+
+        smb.clear();
+        smb.append(iters);
+        if ((iters%pintvl) == 0)
+            PrintLog("MPTEST: MPRing %d send to rank %d", myrank, next);
+        mpicomm->send(smb,next,MPTAG_TEST);
     }
-    else if (numranks > 2)
+
+    PrintLog("MPTEST: MPRing complete");
+
+    mpicomm->barrier();
+
+    delete [] na, pa;
+
+    return;
+}
+
+#define MSGLEN 1048576
+
+void MPAlltoAll(IGroup *group, ICommunicator *_mpicomm, int buffsize=0, int iters=0)
+{
+    int myrank = group->rank();
+    int numranks = group->ordinality();
+
+    if (buffsize == 0)
+        buffsize = MSGLEN;
+    if (iters == 0)
+        iters = 1000;
+    if (iters < 10)
+        iters = 10;
+
+    PrintLog("MPTEST: MPAlltoAll myrank=%d numranks=%d buffsize=%d iters=%d", myrank, numranks, buffsize, iters);
+
+    // ---------
+
+    class Sender : public Thread
     {
-        int m = 0;
-        while (m < (numranks - 1))
+    public:
+        Owned<ICommunicator> mpicomm;
+        int numranks;
+        int myrank;
+        int buffsize;
+        int iters;
+        char *buff;
+        Sender(ICommunicator *_mpicomm, int _numranks, int _myrank, int _buffsize, int _iters)
         {
-            int rankid = 0;
-            while (rankid < numranks)
-            {
-                left = rankid - 1 - m;
-                if (left < 0)
-                    left = numranks + left;
-                right = rankid + 1 + m;
-                if (right >= numranks)
-                    right = right % numranks;
-
-                if (rankid == myrank)
-                {
-                    if (rankid == 0)
-                    {
-                        PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-                        mpicomm->send(mb,right,MPTAG_TEST);
-                        mb2.clear();
-                        PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-                        mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-                        mb2.read(rankval);
-                        ranksum += rankval;
-                    }
-                    else
-                    {
-                        mb2.clear();
-                        PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-                        mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-                        mb2.read(rankval);
-                        ranksum += rankval;
-                        PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-                        mpicomm->send(mb,right,MPTAG_TEST);
-                    }
-                }
-                rankid++;
+            mpicomm.set(_mpicomm);
```

Comment:
more natural to declare member as Linked\<ICommunicator\> mpicomm and then ctor can be:
```
    Sender(ICommunicator *_mpicomm, unsigned _numranks, rank_t _myrank, size32_t _buffsize, unsigned _iters)
        : mpicomm(_mpicomm), numranks(_numranks), myrank(_myrank), iters(_iters)
    {
    }
```